### ISSUE TITLE
Allow `MemberNotNull` on constructors

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#71526]
The current workaround is to define a new property as it is done for `BoundExpression.Type` in roslyn.
```cs
using System.Diagnostics.CodeAnalysis;

new C(1).Value.ToString(); // no warning

class B {
  public B(object? value) => Value = value;
  public object? Value { get; }
}
class C : B {
  [MemberNotNull(nameof(Value))] // currently not allowed 
  public C(object value) : base(value) { }
}
```

Note the member doesn't actually need to be defined in the base so some constructor could be setting this member and some not.

**Comment by sharwell** [comment]
This would need to be changed in the runtime before we could do anything here
https://github.com/dotnet/runtime/blob/1d5f483fa4344328fb85fd67fc482b70b31e6c4d/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs

**Comment by alrz** [comment]
Of course, the compiler team needs to show interest to do this before proposing the change to the runtime.
This alongside of https://github.com/dotnet/roslyn/issues/43906 and possibly others.

**Comment by deepakrathore33** [comment]
https://developercommunity.visualstudio.com/t/Nullable-analysis:-Allow-MemberNotNull-o/11088639
